draw/src/font_atlas.rs

```diff
@@ -286,56 +287,95 @@ impl CxShapeCache {
+                font_atlas,
+                glyph_infos,
+                base_cluster
+            );
         };
 
+        // Create and configure the HarfBuzz buffer.
         let mut buffer = UnicodeBuffer::new();
         buffer.push_str(text);
+
+        // Shape the text using HarfBuzz.
         let buffer = font.owned_font_face.with_ref(|face| {
             makepad_rustybuzz::shape(face, &[], buffer)
         });
 
         let infos = buffer.glyph_infos();
+
+        // Modify the way glyph iteration is handled.
+        let mut skip_to = 0;
+
         let mut info_iter = infos.iter();
-        let mut info_slot = info_iter.next();
-        while let Some(info) = info_slot {
+        while let Some(info) = info_iter.next() {
+            // If this position has already been processed, skip it.
+            if (info.cluster as usize) < skip_to {
+                continue;
+            }
+            // Calculate the absolute cluster position.
+            let absolute_cluster = base_cluster + info.cluster as usize;
+
+            // Handle valid glyphs.
             if info.glyph_id != 0 {
-                info_slot = info_iter.next();
                 glyph_infos.push(GlyphInfo {
                     font_id,
                     glyph_id: info.glyph_id as usize,
-                    cluster: info.cluster as usize,
+                    cluster: absolute_cluster,
                 });
-            } else {
-                let start = info.cluster as usize;
-                let end = loop {
-                    info_slot = info_iter.next();
-                    if let Some(info) = info_slot {
-                        if info.glyph_id != 0 {
-                            break info.cluster as usize;
-                        }
-                    } else {
-                        break text.len()
-                    }
-                };
-                if self.shape_full_recursive(
-                    &text[start..end],
-                    font_ids,
-                    font_atlas,
-                    glyph_infos,
-                ).is_err() {
-                    glyph_infos.push(GlyphInfo {
-                        font_id,
-                        glyph_id: info.glyph_id as usize,
-                        cluster: info.cluster as usize,
-                    });
-                }
+                continue;
+            }
+
+            // Handle missing glyphs.
+            let start = info.cluster as usize;
+
+            // Find the position of the next valid glyph.
+            let next_cluster = {
+                let mut preview_iter = info_iter.clone();
+                let next_valid = preview_iter
+                    .find(|next| next.glyph_id != 0)
+                    .map(|next| next.cluster as usize)
+                    .unwrap_or(text.len());
+                next_valid
+            };
+
+            // Verify the validity of the slice range.
+            if start >= next_cluster {
```

Comment:
Why is this check necessary? Under what circumstances could it fail?

Reply:
This is a defensive check.

Whether this check is still necessary depends on whether we can guarantee that `start` is always less than `next_cluster`.

The generation of cluster values is controlled by the `HarfBuzz` algorithm (`makepad_rustybuzz`). In theory, `HarfBuzz` should guarantee that cluster values are monotonically increasing. 

However, I am not sure if it is always monotonically increasing here, so I made a check as a robust implementation. 

This check is a simple comparison operation with very little overhead, but if we can be sure that this situation will never occur, this check is redundant and can be removed.

Reply:
Or convert this check into a debug assertion, so that potential issues can still be captured during the development phase.
```rust
debug_assert!(
    start < next_cluster,
    "HarfBuzz guarantees monotonic cluster values"
);
```

Reply:
> Or convert this check into a debug assertion, so that potential issues can still be captured during the development phase.或者将此检查转换为调试断言，以便在开发阶段仍然可以捕获潜在的问题。
> 
> ```rust
> debug_assert!(
>     start < next_cluster,
>     "HarfBuzz guarantees monotonic cluster values"
> );
> ```

 I changed it to this debug assertion, and it has been commited.